lib/pack/k8s/KubernetesGoalScheduler.ts

```diff
@@ -231,62 +234,85 @@ export async function cleanCompletedJobs(): Promise<void> {
     }
 }
 
-/**
- * Create a jobSpec by modifying the provided podSpec
- * @param podSpec
- * @param podNs
- * @param gi
- * @param context
- */
-export function createJobSpec(podSpec: k8s.V1Pod, podNs: string, gi: GoalInvocation): k8s.V1Job {
-    const { goalEvent, context } = gi;
-    const goalName = goalEvent.uniqueName.split("#")[0].toLowerCase();
-
-    const jobSpec = createJobSpecWithAffinity(podSpec, gi);
-
-    jobSpec.metadata.name = (`${podSpec.spec.containers[0].name}-job-${goalEvent.goalSetId.slice(0, 7)}-${goalName}`).slice(0, 63);
-    jobSpec.metadata.namespace = podNs;
+/** Unique name for goal to use in k8s job spec. */
+function k8sJobGoalName(goalEvent: SdmGoalEvent): string {
+    return goalEvent.uniqueName.split("#")[0].toLowerCase();
+}
 
-    jobSpec.spec.backoffLimit = 1;
-    jobSpec.spec.template.spec.restartPolicy = "Never";
-    jobSpec.spec.template.spec.containers[0].name = jobSpec.metadata.name;
+/** Unique name for job to use in k8s job spec. */
+export function k8sJobName(podSpec: k8s.V1Pod, goalEvent: SdmGoalEvent): string {
+    const goalName = k8sJobGoalName(goalEvent);
+    return `${podSpec.spec.containers[0].name}-job-${goalEvent.goalSetId.slice(0, 7)}-${goalName}`
+        .slice(0, 63).replace(/[^a-z0-9]*$/, "");
```

Comment:
wouldn't you first replace and then cut to max 63 chars?

Reply:
No. If you don't cut, there will be a lower-case letter or number at the end, so no need to replace. If you cut, you might end up with something other than lower-case letter or number at the end, which is not allowed, so you need the replace. There is a test.